File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/mailbox/v2/MemoryDetailDto.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.data.api.models.mailbox.v2

import app.pantopus.android.data.api.models.common.JsonValue

/**
 * One contextual fact behind the memory ("a year ago today", the
 * originating Pulse thread, where it happened, who else helped).
 */
data class MemoryFact(
    val kind: Kind,
    val label: String,
    val value: String,
    val linkHint: String? = null,
) {
    /** Drives the leading glyph and the row's test-tag. */
    enum class Kind(val raw: String) {
        Anniversary("anniversary"),
        PulseThread("pulseThread"),
        Location("location"),
        Others("others"),
        ;

        companion object {
            fun fromRaw(value: String?): Kind? = entries.firstOrNull { it.raw == value }
        }
    }

    companion object {
        fun decode(map: Map<String, Any?>): MemoryFact? {
            val kind = Kind.fromRaw(map["kind"] as? String) ?: return null
            val label = map["label"] as? String ?: return null
            val value = map["value"] as? String ?: return null
            return MemoryFact(kind, label, value, map["link_hint"] as? String)
        }
    }
}

/** One bullet in the memory elf card. */
data class MemoryElfBullet(
    val glyph: Glyph,
    val label: String,
    val text: String,
) {
    enum class Glyph(val raw: String) {
        Calendar("calendar"),
        Image("image"),
        ShieldCheck("shieldCheck"),
        Archive("archive"),
        EyeOff("eyeOff"),
        Bell("bell"),
        ;

        companion object {
            fun fromRaw(value: String?): Glyph? = entries.firstOrNull { it.raw == value }
        }
    }

    companion object {
        fun decode(map: Map<String, Any?>): MemoryElfBullet? {
            val glyph = Glyph.fromRaw(map["glyph"] as? String) ?: return null
            val label = map["label"] as? String ?: return null
            val text = map["text"] as? String ?: return null
            return MemoryElfBullet(glyph, label, text)
        }
    }
}

/** The "Pantopus surfaced this" elf card. Distinct copy for fresh vs saved. */
data class MemoryElfContent(
    val headline: String,
    val summary: String,
    val bullets: List<MemoryElfBullet>,
) {
    companion object {
        @Suppress("UNCHECKED_CAST")
        fun decode(value: Any?): MemoryElfContent? {
            val map = value as? Map<String, Any?> ?: return null
            val headline = map["headline"] as? String ?: return null
            val summary = map["summary"] as? String ?: return null
            val bullets =
                (map["bullets"] as? List<*>).orEmpty().mapNotNull {
                    (it as? Map<String, Any?>)?.let(MemoryElfBullet::decode)
                }
            return MemoryElfContent(headline, summary, bullets)
        }
    }
}

/** One crumb in the vault-location breadcrumb. */
data class MemoryVaultCrumb(
    val glyph: Glyph,
    val label: String,
    val isCurrent: Boolean,
) {
    enum class Glyph(val raw: String) {
        Inbox("inbox"),
        Archive("archive"),
        Heart("heart"),
        Calendar("calendar"),
        ;

        companion object {
            fun fromRaw(value: String?): Glyph? = entries.firstOrNull { it.raw == value }
        }
    }

    companion object {
        fun decode(map: Map<String, Any?>): MemoryVaultCrumb? {
            val glyph = Glyph.fromRaw(map["glyph"] as? String) ?: return null
            val label = map["label"] as? String ?: return null
            return MemoryVaultCrumb(glyph, label, (map["current"] as? Boolean) ?: false)
        }
    }
}

/** A single counter in the vault summary row. */
data class MemoryVaultStat(
    val value: String,
    val label: String,
) {
    companion object {
        fun decode(map: Map<String, Any?>): MemoryVaultStat? {
            val value = map["value"] as? String ?: return null
            val label = map["label"] as? String ?: return null
            return MemoryVaultStat(value, label)
        }
    }
}

/** Where the memory is filed once kept — replaces the facts grid when saved. */
data class MemoryVaultInfo(
    val trail: List<MemoryVaultCrumb>,
    val stats: List<MemoryVaultStat>,
) {
    companion object {
        @Suppress("UNCHECKED_CAST")
        fun decode(value: Any?): MemoryVaultInfo? {
            val map = value as? Map<String, Any?> ?: return null
            val trail =
                (map["trail"] as? List<*>).orEmpty().mapNotNull {
                    (it as? Map<String, Any?>)?.let(MemoryVaultCrumb::decode)
                }
            if (trail.isEmpty()) return null
            val stats =
                (map["stats"] as? List<*>).orEmpty().mapNotNull {
                    (it as? Map<String, Any?>)?.let(MemoryVaultStat::decode)
                }
            return MemoryVaultInfo(trail, stats)
        }
    }
}

/**
 * Memory-mail sub-payload decoded from `mail.object_payload` when
 * `mail_type == "memory"`. A keepsake delivery: a photograph, a
 * handwritten note, the factual context, and the "why Pantopus surfaced
 * this" elf — plus a vault-location summary for the saved state.
 * [decodeFromObjectPayload] returns null when the payload carries no
 * title or is missing its presentation blocks.
 */
data class MemoryDetailDto(
    val title: String,
    val reference: String,
    val photoUrl: String?,
    val photoCaption: String,
    val photoLabel: String,
    val note: List<String>,
    val noteSignature: String,
    val facts: List<MemoryFact>,
    val elfFresh: MemoryElfContent,
    val elfSaved: MemoryElfContent,
    val vault: MemoryVaultInfo,
    /** True once the user has kept this memory in their vault. */
    val isSaved: Boolean,
) {
    companion object {
        @Suppress("UNCHECKED_CAST", "ReturnCount")
        fun decodeFromObjectPayload(payload: JsonValue?): MemoryDetailDto? {
            if (payload == null) return null
            val title = (payload["title"] as? String)?.takeIf { it.isNotEmpty() } ?: return null
            val elfFresh = MemoryElfContent.decode(payload["elf_fresh"]) ?: return null
            val elfSaved = MemoryElfContent.decode(payload["elf_saved"]) ?: return null
            val vault = MemoryVaultInfo.decode(payload["vault"]) ?: return null

            val photo = payload["photo"] as? Map<String, Any?>
            val note = (payload["note"] as? List<*>).orEmpty().mapNotNull { it as? String }
            val facts =
                (payload["facts"] as? List<*>).orEmpty().mapNotNull {
                    (it as? Map<String, Any?>)?.let(MemoryFact::decode)
                }
            return MemoryDetailDto(
                title = title,
                reference = (payload["reference"] as? String) ?: "",
                photoUrl = photo?.get("url") as? String,
                photoCaption = (photo?.get("caption") as? String) ?: "",
                photoLabel = (photo?.get("label") as? String) ?: "",
                note = note,
                noteSignature = (payload["note_signature"] as? String) ?: "",
                facts = facts,
                elfFresh = elfFresh,
                elfSaved = elfSaved,
                vault = vault,
                isSaved = (payload["is_saved"] as? Boolean) ?: false,
            )
        }
    }
}

File: frontend/apps/ios/Pantopus/Core/Networking/Models/Mailbox/V2/MemoryDetailDTO.swift
//
//  MemoryDetailDTO.swift
//  Pantopus
//
//  Memory-mail sub-payload decoded from `mail.object_payload` when
//  `mail_type == "memory"`. A keepsake delivery: a photograph, a
//  handwritten note, the factual context behind the moment, and the
//  "why Pantopus surfaced this" elf — plus a vault-location summary for
//  the saved state. Backend stores this as untyped JSON in S3 (route
//  handler at `backend/routes/mailboxV2.js:412`); `decode(from:)` is
//  defensive and returns nil when the payload carries no title.
//

import Foundation

/// One contextual fact behind the memory ("a year ago today", the
/// originating Pulse thread, where it happened, who else helped).
public struct MemoryFact: Sendable, Hashable, Identifiable {
    /// Drives both the leading glyph and the row's `accessibilityIdentifier`.
    public enum Kind: String, Sendable, Hashable {
        case anniversary
        case pulseThread
        case location
        case others
    }

    public let kind: Kind
    public let label: String
    public let value: String
    /// Optional link affordance ("tap to reopen the thread"). When set the
    /// row is rendered as a tappable button.
    public let linkHint: String?

    public var id: String {
        kind.rawValue
    }

    public init(kind: Kind, label: String, value: String, linkHint: String? = nil) {
        self.kind = kind
        self.label = label
        self.value = value
        self.linkHint = linkHint
    }

    static func decode(from value: JSONValue) -> MemoryFact? {
        guard let dict = value.dictValue,
              let rawKind = dict["kind"]?.stringValue,
              let kind = Kind(rawValue: rawKind),
              let label = dict["label"]?.stringValue,
              let factValue = dict["value"]?.stringValue
        else { return nil }
        return MemoryFact(
            kind: kind,
            label: label,
            value: factValue,
            linkHint: dict["link_hint"]?.stringValue
        )
    }
}

/// One bullet in the memory elf card.
public struct MemoryElfBullet: Sendable, Hashable, Identifiable {
    public enum Glyph: String, Sendable, Hashable {
        case calendar
        case image
        case shieldCheck
        case archive
        case eyeOff
        case bell
    }

    public let glyph: Glyph
    public let label: String
    public let text: String

    public var id: String {
        label
    }

    public init(glyph: Glyph, label: String, text: String) {
        self.glyph = glyph
        self.label = label
        self.text = text
    }

    static func decode(from value: JSONValue) -> MemoryElfBullet? {
        guard let dict = value.dictValue,
              let rawGlyph = dict["glyph"]?.stringValue,
              let glyph = Glyph(rawValue: rawGlyph),
              let label = dict["label"]?.stringValue,
              let text = dict["text"]?.stringValue
        else { return nil }
        return MemoryElfBullet(glyph: glyph, label: label, text: text)
    }
}

/// The "Pantopus surfaced this" elf card. Distinct copy for fresh vs
/// saved states.
public struct MemoryElfContent: Sendable, Hashable {
    public let headline: String
    public let summary: String
    public let bullets: [MemoryElfBullet]

    public init(headline: String, summary: String, bullets: [MemoryElfBullet]) {
        self.headline = headline
        self.summary = summary
        self.bullets = bullets
    }

    static func decode(from value: JSONValue?) -> MemoryElfContent? {
        guard let dict = value?.dictValue,
              let headline = dict["headline"]?.stringValue,
              let summary = dict["summary"]?.stringValue
        else { return nil }
        let bullets = (dict["bullets"]?.arrayValue ?? []).compactMap(MemoryElfBullet.decode(from:))
        return MemoryElfContent(headline: headline, summary: summary, bullets: bullets)
    }
}

/// One crumb in the vault-location breadcrumb.
public struct MemoryVaultCrumb: Sendable, Hashable, Identifiable {
    public enum Glyph: String, Sendable, Hashable {
        case inbox
        case archive
        case heart
        case calendar
    }

    public let glyph: Glyph
    public let label: String
    public let isCurrent: Bool

    public var id: String {
        label
    }

    public init(glyph: Glyph, label: String, isCurrent: Bool) {
        self.glyph = glyph
        self.label = label
        self.isCurrent = isCurrent
    }

    static func decode(from value: JSONValue) -> MemoryVaultCrumb? {
        guard let dict = value.dictValue,
              let rawGlyph = dict["glyph"]?.stringValue,
              let glyph = Glyph(rawValue: rawGlyph),
              let label = dict["label"]?.stringValue
        else { return nil }
        return MemoryVaultCrumb(
            glyph: glyph,
            label: label,
            isCurrent: dict["current"]?.boolValue ?? false
        )
    }
}

/// A single counter in the vault summary row.
public struct MemoryVaultStat: Sendable, Hashable, Identifiable {
    public let value: String
    public let label: String

    public var id: String {
        label
    }

    public init(value: String, label: String) {
        self.value = value
        self.label = label
    }

    static func decode(from value: JSONValue) -> MemoryVaultStat? {
        guard let dict = value.dictValue,
              let statValue = dict["value"]?.stringValue,
              let label = dict["label"]?.stringValue
        else { return nil }
        return MemoryVaultStat(value: statValue, label: label)
    }
}

/// Where the memory is filed once kept — rendered in the saved state in
/// place of the facts grid.
public struct MemoryVaultInfo: Sendable, Hashable {
    public let trail: [MemoryVaultCrumb]
    public let stats: [MemoryVaultStat]

    public init(trail: [MemoryVaultCrumb], stats: [MemoryVaultStat]) {
        self.trail = trail
        self.stats = stats
    }

    static func decode(from value: JSONValue?) -> MemoryVaultInfo? {
        guard let dict = value?.dictValue else { return nil }
        let trail = (dict["trail"]?.arrayValue ?? []).compactMap(MemoryVaultCrumb.decode(from:))
        let stats = (dict["stats"]?.arrayValue ?? []).compactMap(MemoryVaultStat.decode(from:))
        guard !trail.isEmpty else { return nil }
        return MemoryVaultInfo(trail: trail, stats: stats)
    }
}

/// Memory detail payload — drives the A17.7 keepsake body.
public struct MemoryDetailDTO: Sendable, Hashable {
    public let title: String
    public let reference: String
    public let photoURL: URL?
    public let photoCaption: String
    public let photoLabel: String
    public let note: [String]
    public let noteSignature: String
    public let facts: [MemoryFact]
    public let elfFresh: MemoryElfContent
    public let elfSaved: MemoryElfContent
    public let vault: MemoryVaultInfo
    /// True once the user has kept this memory in their vault.
    public let isSaved: Bool

    public init(
        title: String,
        reference: String,
        photoURL: URL?,
        photoCaption: String,
        photoLabel: String,
        note: [String],
        noteSignature: String,
        facts: [MemoryFact],
        elfFresh: MemoryElfContent,
        elfSaved: MemoryElfContent,
        vault: MemoryVaultInfo,
        isSaved: Bool
    ) {
        self.title = title
        self.reference = reference
        self.photoURL = photoURL
        self.photoCaption = photoCaption
        self.photoLabel = photoLabel
        self.note = note
        self.noteSignature = noteSignature
        self.facts = facts
        self.elfFresh = elfFresh
        self.elfSaved = elfSaved
        self.vault = vault
        self.isSaved = isSaved
    }

    /// Returns a copy with `isSaved` flipped — used by the view-model's
    /// optimistic "save to vault" flow.
    public func withSaved(_ saved: Bool) -> MemoryDetailDTO {
        MemoryDetailDTO(
            title: title,
            reference: reference,
            photoURL: photoURL,
            photoCaption: photoCaption,
            photoLabel: photoLabel,
            note: note,
            noteSignature: noteSignature,
            facts: facts,
            elfFresh: elfFresh,
            elfSaved: elfSaved,
            vault: vault,
            isSaved: saved
        )
    }

    /// Best-effort decode. Returns nil when the payload is missing the
    /// bare-minimum field set (`title`) or its presentation blocks
    /// (`elf_fresh`, `elf_saved`, `vault`).
    public static func decode(from value: JSONValue?) -> MemoryDetailDTO? {
        guard let dict = value?.dictValue else { return nil }
        guard let title = dict["title"]?.stringValue, !title.isEmpty else { return nil }
        guard let elfFresh = MemoryElfContent.decode(from: dict["elf_fresh"]),
              let elfSaved = MemoryElfContent.decode(from: dict["elf_saved"]),
              let vault = MemoryVaultInfo.decode(from: dict["vault"])
        else { return nil }

        let photo = dict["photo"]?.dictValue
        let note = (dict["note"]?.arrayValue ?? []).compactMap(\.stringValue)
        let facts = (dict["facts"]?.arrayValue ?? []).compactMap(MemoryFact.decode(from:))

        return MemoryDetailDTO(
            title: title,
            reference: dict["reference"]?.stringValue ?? "",
            photoURL: photo?["url"]?.stringValue.flatMap(URL.init(string:)),
            photoCaption: photo?["caption"]?.stringValue ?? "",
            photoLabel: photo?["label"]?.stringValue ?? "",
            note: note,
            noteSignature: dict["note_signature"]?.stringValue ?? "",
            facts: facts,
            elfFresh: elfFresh,
            elfSaved: elfSaved,
            vault: vault,
            isSaved: dict["is_saved"]?.boolValue ?? false
        )
    }
}
